cpp/src/arrow/util/thread_pool.h

```diff
@@ -373,7 +377,7 @@ class ARROW_EXPORT ThreadPool : public Executor {
   State* state_;
   bool shutdown_on_destroy_;
 #ifndef _WIN32
-  pid_t pid_;
+  std::atomic<pid_t> pid_;
```

Comment:
Does `pid_` need to be atomic since you're double checking and the second check is inside a mutex?

Reply:
That should be ok because we have the double check on line 247 which happens inside the mutex.

Reply:
Agreed the code should be safe in practice.
But at least in theory, I think it still violates data race free rule.
E.g., TSAN will complain below code.
```
#include <mutex>
#include <thread>

std::mutex g_mtx;
int g_i = 0;

void f() {
  if (g_i == 0) {
    std::lock_guard<std::mutex> lk(g_mtx);
    if (g_i == 0) {
      g_i = 1;
    }
  }
}

int main(void) {
  std::thread t1(f), t2(f), t3(f), t4(f);
  t1.join(); t2.join(); t3.join(); t4.join();
  return 0;
}
```

Reply:
Yes, an atomic is required at least for theoretical correctness and to avoid (potentially) tripping up TSAN.
The compiler and/or the CPU is otherwise theoretically allowed to issue the integer read as multiple distinct reads (say, issue a 32-bit read as two 16-bit reads; I suppose that never happens in practice :-)

Reply:
Just to make this discussion more complete.

If we only care about atomicity, to satisfy TSAN without introducing unnecessary memory barriers, we can use relaxed memory ordering when accessing the atomic variable, not the default sequential consistent memory ordering.

The assembly code will probably be the same as a normal variable on most architectures.
```
if (g_i.load(std::memory_order_relaxed) == 0) {
  std::lock_guard<std::mutex> lk(g_mtx);
  if (g_i.load(std::memory_order_relaxed) == 0) {
    g_i.store(1, std::memory_order_relaxed);
  }
}
```

This might help performance on cpus with more relaxed memory models than x86 (e.g., arm, ppc) in some cases, but the improvement is often negligible in a real world application.

Reply:
Thank you for the details @cyb70289 .  I appreciate you both taking the time to help explain this.   This code shouldn't be on the data path so I agree the extra performance is not worth false positives in TSAN or the extra complexity of relaxed memory order.